File: Ballast.Core/src/Server/Models/OffsetCoordinates.cs
using System;

namespace Ballast.Core.Models
{

    public class OffsetCoordinatesState
    {
        public int Row { get; set; }
        public int Col { get; set; }
    }

    public class OffsetCoordinates
    {

        public int Row { get; private set; }
        public int Col { get; private set; }

        private OffsetCoordinates(int col, int row) 
        {
            Col = col;
            Row = row;        
        }

        public static OffsetCoordinates FromProperties(int col, int row) => 
            new OffsetCoordinates(
                col: col,
                row: row
            );

        public static OffsetCoordinates FromAxial(AxialCoordinates axial) 
        {
            // Bitwise AND (& 1) to get 0 for even or 1 for odd column offset
            var col = axial.X + (axial.Z - (axial.Z & 1)) / 2;
            var row = axial.Z;
            return new OffsetCoordinates(col: col, row: row);
        }
        
        public static OffsetCoordinates FromCubic(CubicCoordinates cubic) =>
            OffsetCoordinates.FromAxial(
                AxialCoordinates.FromCubic(cubic)
            );
        
        public static OffsetCoordinates FromOrderedPair(int[] orderedPair) {
            if (orderedPair.Length < 2)
                throw new Exception("Length of ordered pair must be 2 (or greater)");
                var col = orderedPair[0]; // x
                var row = orderedPair[1]; // z
            return new OffsetCoordinates(col: col, row: row);
        }

        public static implicit operator OffsetCoordinates(OffsetCoordinatesState state) =>
            new OffsetCoordinates(state.Col, state.Row);

        public AxialCoordinates ToAxial() => AxialCoordinates.FromOffset(this);

        public CubicCoordinates ToCubic() => CubicCoordinates.FromOffset(this);

        public int[] ToOrderedPair() => new int[] { Col, Row };

    }
    
}

File: ballast-core/npm/src/domain/models/offset-coordinates.ts
import { AxialCoordinates } from './axial-coordinates';
import { CubicCoordinates } from './cubic-coordinates';

export class OffsetCoordinates {

    public readonly col: number;
    public readonly row: number;

    private constructor(col: number, row: number) {
        this.col = col;
        this.row = row;        
        if (this.col == -0)
            this.col = 0;
        if (this.row == -0)
            this.row = 0;
    }

    public static fromAxial(object: AxialCoordinates) {
        // Bitwise AND (& 1) to get 0 for even or 1 for odd column offset
        let col = object.x + (object.z - (object.z & 1)) / 2;
        let row = object.z;
        return new OffsetCoordinates(col, row);
    }

    public static fromCubic(object: CubicCoordinates) {
        // Bitwise AND (& 1) to get 0 for even or 1 for odd column offset
        let col = object.x + (object.z - (object.z & 1)) / 2;
        let row = object.z;
        return new OffsetCoordinates(col, row);
    }

    public static fromOrderedPair(orderedPair: number[]) {
        if (orderedPair.length < 2)
            throw new Error('Length of ordered pair must be 2 (or greater)');
            let col = orderedPair[0]; // x
            let row = orderedPair[1]; // z
        return new OffsetCoordinates(col, row);
    }

    public toAxial() {
        return AxialCoordinates.fromOffset(this);
    }

    public toCubic() {
        return CubicCoordinates.fromOffset(this);
    }

    public toOrderedPair() {
        return [this.col, this.row];
    }

}